frontend/server/utils.ts

```diff
@@ -65,6 +66,23 @@ export function loadJSON<T>(filepath?: string, defaultValue?: T): T | undefined
   }
 }
 
+/**
+ * prune subMaybePrunedPath to relativePath, then concat with rootFixedPath.
+ */
+export function pruneAndConcatPath(
+  rootFixedPath: string,
+  subMaybePrunedPath: string,
+  prefixPrune: string | undefined,
+): string {
+  if (!prefixPrune) {
+    return path.join(rootFixedPath, subMaybePrunedPath);
+  }
+  if (subMaybePrunedPath.startsWith(prefixPrune)) {
+    return path.join(rootFixedPath, subMaybePrunedPath.substring(prefixPrune.length));
+  }
+  throw new Error(`${prefixPrune} should be prefix of ${subMaybePrunedPath}`);
```

Comment:
continuing on above comment, when we have more context in the function signature, we can also make this error message easier to understand.

probably `throw new Error(`File ${filePathInVolume} not mounted, expecting the file to be inside volume mount subpath ${volumeMountSubPath}.`)`.